hooks/pre-checkout

```diff
@@ -3,56 +3,123 @@ set -eo pipefail
 
 DESTINATION=/tmp/$BUILDKITE_PIPELINE_SLUG.git.tar
 
-# If we have an available Git Mirror Server, use it
-if [ -n "${GIT_MIRROR_SERVER_ROOT:-}" ]; then
-	echo "Using Git Mirror Server at $GIT_MIRROR_SERVER_ROOT"
-	URL=$(curl "$GIT_MIRROR_SERVER_ROOT/manifest" | grep $REPO_PATH)
+git_mirror_download () {
+  echo "Using Git Mirror Server at $GIT_MIRROR_SERVER_ROOT"
+	if URL=$(curl "$GIT_MIRROR_SERVER_ROOT/manifest" | grep $REPO_PATH$); then
+    echo "Downloading snapshot $URL to $DESTINATION"
+    curl -so "$DESTINATION" "$GIT_MIRROR_SERVER_ROOT/$URL"
+    echo ":large_green_circle: Cache download successful!"
+    return 0
+  else
+    echo "  ↳ WARNING: No snapshots found in $GIT_MIRROR_SERVER_ROOT. Will attempt s3."
+    ERRORS_ARRAY+=("- WARNING: No snapshots found in $GIT_MIRROR_SERVER_ROOT. Will attempt s3.")
+		return 1
+  fi
+}
 
-	echo "Downloading snapshot $URL to $DESTINATION"
-	curl -so "$DESTINATION" "$GIT_MIRROR_SERVER_ROOT/$URL"
-# Otherwise, use S3 directly
-else
-	# Fetch the most recent S3 backup key
-	SNAPSHOT_KEY=$(aws --output json s3api list-objects-v2 --bucket $BUCKET --prefix $REPO_PATH --query 'Contents[].Key' | jq -r '.[] | select(endswith(".git.tar"))' | sort -r | head -n1)
+s3_download () {
+  # Fetch the most recent S3 backup key
+	SNAPSHOT_KEY=$(aws --output json s3api list-objects-v2 --bucket "$BUCKET" --prefix "$REPO_PATH" --query 'Contents[].Key' | jq -sr 'sort[] | reverse[] | select(endswith(".git.tar"))' | head -n1 )
 
-	# Exit early if there's no available snapshot
+  # Exit early if there's no available snapshot
 	if [ -z "$SNAPSHOT_KEY" ]; then
-		echo "No snapshots found in $BUCKET"
-		exit 0
+		echo "  ↳ WARNING: No snapshots found in $BUCKET"
+    ERRORS_ARRAY+=("- WARNING: No snapshots found in $BUCKET")
+		return 1
 	fi
 
-	echo "Downloading snapshot: $SNAPSHOT_KEY to $DESTINATION"
+ SNAPSHOT_KEY_CHECK=$(echo "$SNAPSHOT_KEY" | cut -d "/" -f 1-2)
+  if [ "$SNAPSHOT_KEY_CHECK" == "$REPO_PATH" ]; then
+		echo "Downloading snapshot: $SNAPSHOT_KEY to $DESTINATION"
+    # Then download it
+    rm -rf "$DESTINATION" # Delete before starting, just in case
+    aws s3 cp "s3://$BUCKET/$SNAPSHOT_KEY" "$DESTINATION"
+    echo ":large_green_circle: S3 Download Succeeded"
+    return 0
+  else
+    echo "  ↳ WARNING: Nothing downloaded. $SNAPSHOT_KEY is not the same as $REPO_PATH (comparison used $SNAPSHOT_KEY_CHECK)"
+    ERRORS_ARRAY+=("- WARNING: Nothing downloaded. $SNAPSHOT_KEY is not the same as $REPO_PATH")
+    return 1
+  fi
+}
+
+annotate_error() {
+  ERRORS_ARRAY=("$@")
+  ERRORS=$(
+    for error in "${ERRORS_ARRAY[@]}"
+    do
+      echo "$error"
+    done
+  )
 
-	# Then download it
-	rm -rf $DESTINATION # Delete before starting, just in case
-	aws s3 cp "s3://$BUCKET/$SNAPSHOT_KEY" "$DESTINATION"
+  buildkite-agent annotate --style "warning" --context "warnings" --append \
+"
+Step **$BUILDKITE_LABEL** had _WARNINGS_!
+Agent: **$BUILDKITE_AGENT_NAME** <br>
```

Comment:
The label and agent name variables currently fail in the Mac VMs because they are not transferred.  [This PR address that issue.](https://github.com/Automattic/hostmgr/pull/18)